src/formatter.rs

```diff
@@ -1,108 +1,230 @@
-    example_code: &'a str,
-    config: &Config,
-    parts: &mut Vec<ANSIString<'a>>,
-) {
-    let mut code_part_end_pos = 0;
-    while let Some(command_start) = example_code[code_part_end_pos..].find(&command) {
-        let code_part = &example_code[code_part_end_pos..code_part_end_pos + command_start];
-        parts.push(config.style.example_code.paint(code_part));
-        if code_part_end_pos == 0 {
-            // Only highlight command names at the start of the line ...
-            parts.push(config.style.command_name.paint(command));
-        } else {
-            let char_before_command = example_code
-                .chars()
-                .nth(code_part_end_pos + command_start - 1);
-            if char_before_command.filter(|c| c.is_whitespace()).is_some() {
-                // ... or when preceded by a whitespace character.
-                parts.push(config.style.command_name.paint(command));
-            } else {
-                parts.push(config.style.example_code.paint(command));
-            }
-        }
+use log::debug;
 
-        code_part_end_pos += command_start + command.len();
-    }
-    parts.push(
-        config
-            .style
-            .example_code
-            .paint(&example_code[code_part_end_pos..]),
-    );
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+/// Represents a snippet from a page from a specific highlighting class.
+pub enum HighlightingSnippet<'a> {
+    CommandName(&'a str),
+    Variable(&'a str),
+    NormalCode(&'a str),
+    Description(&'a str),
+    Text(&'a str),
+    Linebreak,
 }
 
-/// Format and highlight code examples including variables in {{ curly braces }}.
-fn format_code(command: &str, text: &str, config: &Config) -> String {
-    let mut parts = Vec::new();
-    for between_variables in text.split("}}") {
-        if let Some(variable_start) = between_variables.find("{{") {
-            let example_code = &between_variables[..variable_start];
-            let example_variable = &between_variables[variable_start + 2..];
-
-            highlight_command(&command, &example_code, &config, &mut parts);
-            parts.push(config.style.example_variable.paint(example_variable));
-        } else {
-            highlight_command(&command, &between_variables, &config, &mut parts);
+impl<'a> HighlightingSnippet<'a> {
+    pub fn is_empty(&self) -> bool {
+        use HighlightingSnippet::*;
+
+        match self {
+            CommandName(s) | Variable(s) | NormalCode(s) | Description(s) | Text(s) => s.is_empty(),
+            Linebreak => false,
         }
     }
-
-    ANSIStrings(&parts).to_string()
 }
 
-/// Print a token stream to an ANSI terminal.
-pub fn print_lines<T, R>(
-    writer: &mut T,
-    tokenizer: &mut Tokenizer<R>,
-    config: &Config,
-) -> Result<(), TealdeerError>
+/// Parse the content of each line yielded by `lines` and yield `HighLightingSnippet`s accordingly.
+pub fn highlight_lines<L, F, E>(
+    lines: L,
+    yield_snippet: &mut F,
```

Comment:
The `yield` terminology confused me a bit, I know it mostly from the context of iterators/generators (it's fine for `lines`, which _is_ an iterator).

What do you think about renaming the closure (and related code / comments) to `process_snippet` or `emit_snippet`?

Reply:
Ah that's unlucky, I thought it would make it easier to understand if these function actually followed iterator naming conventions, because they really are an iterator adapter or rather a stream adapter. I would really like to model this out as a big stream of `*Snippet`, but it will probably get difficult with lifetimes. The current implementation requires callers to give the generic handler `yield_snippet` to immediately dispose of any emitted snippets instead of accumulating them somewhere (like a more abstract stream model would).

I can see that the responsibilities taken by the different modules are a bit weird, maybe an approach that parses code lines before even yielding the line would be cleaner, we could explore that route

Reply:
I see. With the current code, I still think it's better to use `process_snippet` or `emit_snippet`. Pick whichever you prefer (I think I'd prefer "process").